Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp

```diff
@@ -182,20 +182,36 @@ ECMAScriptFunctionObject::ECMAScriptFunctionObject(DeprecatedFlyString name, Dep
         m_arguments_object_needed = false;
     }
 
+    size_t* environment_size = &m_function_environment_bindings_count;
+
+    Optional<size_t> parameter_environment_bindings_count;
+    if (m_strict || !m_has_parameter_expressions) {
+        environment_size = &m_function_environment_bindings_count;
+    } else {
+        parameter_environment_bindings_count = 0;
+        environment_size = &parameter_environment_bindings_count.value();
+    }
```

Comment:
This code (and subsequent repetitions of the same idea) is very hard to read.

```cpp
Optional<size_t> parameter_environment_bindings_count;
```

Why does this need to be `Optional`?

```cpp
if (m_strict || !m_has_parameter_expressions) {
    environment_size = &m_function_environment_bindings_count;
}
```

`environment_size` is already pointing at `m_function_environment_bindings_count` so this is a no op.

```cpp
    parameter_environment_bindings_count = 0;
    environment_size = &parameter_environment_bindings_count.value();
```

Why do we want to reroute `environment_size` so it points at a temporary here? I don't understand the logic, and there's no comment explaining what we're trying to achieve :)

Reply:
here I wanted to follow what the spec says: if separate environment needs to be allocated for parameters then we also need separate variable that will store its size.

I added spec comments to make it look less confusing.